backend/src/main/java/fk/prof/backend/model/assignment/WorkAssignmentSchedule.java

```diff
@@ -0,0 +1,211 @@
+                                int minAcceptableDelayForWorkAssignmentInSecs,
+                                int maxAcceptableDelayForWorkAssignmentInSecs,
+                                Recorder.WorkAssignment.Builder[] workAssignmentBuilders,
+                                int coveragePct,
+                                int maxConcurrentSlotsAllowed,
+                                int profileDurationInSecs)
+      throws IllegalArgumentException {
+
+    // breathing space at the start, this will usually be a lower value that window tolerance
+    int windowStartToleranceInSecs = schedulingBufferInSecs * 2;
+    //actual time span for which schedule is calculated
+    int effectiveWindowDurationInSecs = (windowDurationInMins * 60) - windowEndToleranceInSecs - windowStartToleranceInSecs;
+    int effectiveProfileDurationInSecs = profileDurationInSecs + schedulingBufferInSecs;
+    int maxScheduleEntriesWithNoOverlap = effectiveWindowDurationInSecs / effectiveProfileDurationInSecs;
+    int maxScheduleEntriesWithOverlap = maxScheduleEntriesWithNoOverlap * maxConcurrentSlotsAllowed;
+    int targetScheduleEntries = workAssignmentBuilders.length;
+    if (maxScheduleEntriesWithOverlap < targetScheduleEntries) {
+      throw new IllegalArgumentException(String.format("Not possible to setup schedule for " +
+          "vm_coverage=%d, target_profiles=%d, profile_duration_secs=%d, max concurrent slots=%d",
+          coveragePct, targetScheduleEntries, profileDurationInSecs, maxConcurrentSlotsAllowed));
+    }
+
+    this.minAcceptableDelayForWorkAssignmentInSecs = minAcceptableDelayForWorkAssignmentInSecs;
+    this.maxAcceptableDelayForWorkAssignmentInSecs = maxAcceptableDelayForWorkAssignmentInSecs;
+    if(this.maxAcceptableDelayForWorkAssignmentInSecs < (this.minAcceptableDelayForWorkAssignmentInSecs * 2)) {
+      throw new IllegalArgumentException(String.format("Max acceptable delay for work assignment = %d" +
+          "should be at least be twice of min acceptable delay for work assignment = %d", maxAcceptableDelayForWorkAssignmentInSecs, minAcceptableDelayForWorkAssignmentInSecs));
+    }
+
+    this.referenceTimeInNanos = System.nanoTime();
+    long initialOffsetInNanos = windowStartToleranceInSecs * NANOS_IN_SEC;
+    int minOverlap = targetScheduleEntries / maxScheduleEntriesWithNoOverlap;
+
+    int currentWorkAssignmentIndex = 0;
+    for (int i = 0; (minOverlap > 0) && (i < maxScheduleEntriesWithNoOverlap); i++) {
+      long offsetInNanos = initialOffsetInNanos + (effectiveProfileDurationInSecs * i * NANOS_IN_SEC);
+      for(int j = 0; j < minOverlap; j++) {
+        this.entries.add(new ScheduleEntry(workAssignmentBuilders[currentWorkAssignmentIndex++], offsetInNanos));
+      }
+    }
+
+    int remainingAssignments = targetScheduleEntries % maxScheduleEntriesWithNoOverlap;
+    assert targetScheduleEntries == (currentWorkAssignmentIndex + remainingAssignments);
+    for (int i = 0; i < remainingAssignments; i++) {
```

Comment:
Fixed in commit https://github.com/Flipkart/fk-prof/commit/eaa57eb6a8bb821fd8543e80300e2965bbafcb28